### PR TITLE
[PR 496 rework] Disable USART IRQ in uart_write and uart_debug_write 

### PR DESCRIPTION
**Pull request by fpistm**
This PR is a rework of #496  and fixes issues #494 and #467 

It hardens and enhances the original PR.
I've tested several config and it works with Arduino Serial monitor.
Note that if the same U(S)ART is used for `Serial` and `printf()` some bytes could be lost due to IRQ disabling.

@ppescher  and @benwaffle could you review and test it on your side

### CODE REVIEW

File: cores/arduino/stm32/uart.c [hunk 1]
```diff
@@ -581,38 +581,45 @@ void uart_debug_init(void)
...
     if ((HAL_GetTick() - tickstart) >=  TX_TIMEOUT) {
-      return 0;
+      size = 0;
+      break;
     }
   }
 
+  serial_t *obj = rx_callback_obj[serial_debug.index];
+  if (obj && !serial_rx_active(obj)) {
+    HAL_UART_Receive_IT(uart_handlers[obj->index], &(obj->recv), 1);
```
Comment (benwaffle) [inline cores/arduino/stm32/uart.c]: Why are you calling HAL_UART_Receive_IT in uart_debug_write?

**Comment by benwaffle** [comment]
> Note that if the same U(S)ART is used for Serial and printf() some bytes could be lost due to IRQ disabling.
Can you explain under what conditions this would happen?

File: cores/arduino/stm32/uart.c (hunk 1 above)
Comment (fpistm) [reply cores/arduino/stm32/uart.c]: In fact, I've added this to test. It is removed.
I've fixed an issue with the IRQ which was not correct if Serial already init.

**Comment by fpistm** [comment]
> 
> 
> > Note that if the same U(S)ART is used for Serial and printf() some bytes could be lost due to IRQ disabling.
> 
> Can you explain under what conditions this would happen?

In fact by default, the `printf` will send on the same `U(S)ART` used for the generic instance.
In this case, when the first `Serial.available` is true, then `printf` is called and disabled the `U(S)ART` interrupt then in this window the data is simply lost until the interrupt was set.

**Comment by fpistm** [comment]
I've added some code enhancement which allow to save space and avoid loop to search the right `serial_t obj`.

**Pull request merged**